# Materia oscura, Curvas de rotación de Galaxias y Estadística Bayesiana

Por: Iker Alejandro Aldecoa Rendón.

## Resumen

En este notebook se aborda el concepto de materia oscura, la evidencia observacional relacionado a ella, se profundiza específicamente en el tópico de las curvas de rotación de las galaxias, la aplicación que tienen los algoritmos MCMC en esta área y se desarrollan dos ejemplos de implementación de EMCEE utilizando conjuntos de datos reales obtenidos del proyecto SPARC.

El propósito en general es que sirva como una herramienta para aquellos estudiantes de de nivel medio superior o primeros semestre de licenciatura en física o matemáticas que quisieran conocer un poco más sobre estos tópicos, el estado actual del problema, las bases de datos que existen y cómo se puede trabajar con esos datos utilizando materiales que les sean accesibles si conocen un poco de programación en Python, así como las alternativas que existen en R y Julia.

## Tabla de Contenido
1. Materia Oscura.
   1. ¿Qué es la materia oscura?
   2. Evidencia observacional.
   3. Hipótesis alternativas y estado actual.
2. Curvas de rotación.
   1. Lo que esperábamos y lo que observamos.
   2. Estado actual.
   3. Aplicación de la estadística Bayesiana.
      1. Métodos MCMC y algoritmo Goodman & Weare.
3. Ejemplo de aplicación.
   1. Librerías.
   2. Los Datos.
   3. El Modelo.
   4. El Prior.
   5. Construir la función Chi-Cuadrada.
   6. Likelihood.
   7. La Posterior.
   8. Walkers y EMCEE.
   9. Burn-in y Resultados.
   10. Reutilizar EMCEE.
4. Conclusiones.
5. Cuestionario.
6. Bibliografía.


## 1. Materia Oscura

### 1.A ¿Qué es la materia oscura?

Es muy probable que, incluso sin haber estudiado nada relacionado directamente con cosmología, en algún momento hayamos escuchado hablar de la "materia oscura" en algún medio de entretenimiento como videojuegos, series, comics, etc. Ya sea que haya sido de esta forma, por algún comentario en clase, alguna charla o vídeo divulgativo u otras fuentes seguramente lo primero que se nos vino a la cabeza fue la pregunta ¿Qué es la materia oscura? De forma breve la respuesta sería: no sabemos. Sin embargo, si quisiéramos ahondar un poco más, una respuesta más completa sería que, aunque no sabemos *¿Qué?* es, si sabemos *¿Cómo?* se comporta.

Desde que fue formulada por primera vez en 1933 por Fritz Zwicky [1], el tópico de la materia oscura es uno de los mayores retos de la cosmología y la astrofísica moderna [1]. En general lo entendemos como "una forma hipotética de materia que no emite suficiente radiación electromagnética como para ser detectada con los medios técnicos actuales" [2]. Esto quiere decir que la materia oscura no emite luz (de ahí su nombre), no absorbe ningún tipo de radiación de forma apreciable y no interactúa electromagéticamente de forma detectable [3]. Al resto de materia (la "común" pudiéramos decir) se le denomina "materia bariónica".
Uno pudiera pensar que un objeto que suena tan *exótico* en realidad sería raro, pero la realidad es que no. Según el modelo $\Lambda$CDM (el modelo cosmológico estándar) esta consituye cerca del 27% del contenido total de masa-energía del universo y al rededor del 85% de toda la materia [4, 5] **¡Eso es más de 5 veces la materia bariónica!** Por lo que podríamos decir que, a escala cosmológica, estamos rodeados por materia oscura. 

![Proporción de materia oscura](Imagenes/Planck_cosmic_recipe.jpg)

**Figura 1.** Proporción de materia oscura, materia bariónica y energía oscura en el universo.

**Fuente:** ESA (2013), *Planck Cosmic Recipe*.

Además de su abundancia y su falta de interacción electromagnética, también se sabe que es fría, es escencialmente no colisional, sus partículas interaccionan muy débilmente y debió existir desde los momentos iniciales del big bang aunque aún desconocemos su composición microscópica [1, 2]. Todas estas propiedades se obtienen a partir de la evidencia observacional de esta (véase la siguiente sección).

La importancia de entender la materia oscura radica en su relación con problemas como la composición del universo, la formación de las galaxias y otras estructuras cosmológicas, la teoría de la gravitación y la relatividad [2, 6, 7].

### 1.B Evidencia Observacional

Como tal vez te habrás imaginado, el hecho de que tenga muy pocas interacciones hace que sea difícil poder detectar la materia oscura [8, 9]. Sin embargo, sabemos que esta sí interacciona gravitacionalmente y es justamente esto lo que nos permite conocer un poco más sobre ella [1]; además de que es también uno de los motivos por el que es una pieza clave en los problemas que mencionamos anteriormente [2].

Algunas de las evidencias observacionales más importantes son:
* Curvas de Rotación: La velocidad de las estrellas y gases en galaxias espriales se mantiene prácticamente constante a grandes distancias del centro de la galaxia, algo que no tiene sentido si sólo consideramos la materia visible [2,10].

![Curvas de Rotación](Imagenes/Curva_De_Rot1.png)

**Figura 2.** Curva de rotación de la galaxia que muestra la diferencia entre la curva observada (línea verde) y la curva esperada (línea amarilla).

**Accesible en:** https://www.ufrn.br/es/press/features-and-knowledge/60434/solitario-e-a-deriva-pelo-espaco

* Dinámica de Cúmulos de Galaxias: Esta fue la primer evidencia moderna, obtenida por Fritz Zwicky [11]. Él encontró que las galaxias del cúmulo se movían demasiado rápido para que se mantuviera unido únicamente con la materia visible, lo que indicaba la presencia de algún tipo de materia oscura [2].

![Cúmulo de galaxias ACT-CL J0102-4915](Imagenes/eso1203a.jpg)

**Figura 3.** Fotografía del cúmulo de galaxias ACT-CL J0102−4915.

**Fuente:** ESO (2012), *eso1203a*.

* Lentes Gravitacionales: Al estudiar los lentes gravitacionales, la luz de las galaxias de fondo se curva más de lo previsto considerando solo la materia bariónica [12].

![Diagrama de un lente gravitacional.](Imagenes/eso1426b.jpg)

**Figura 4.** Diagrama de un lente gravitacional.

**Fuente:** ESO (2014), *eso1426b*.

* Fondo Cósmico de Microondas: Las pequeñas variaciones de temperatura y densidad observadas en el Fondo Cósmico de Microondas (CMB) indica la presencia de materia oscura para la época de recombinación [13].

![Anisotropía en el fondo cósmico de microondas.](Imagenes/Anisotropia.png)

**Figura 5.** Anisotropía en el fondo cósmico de microondas.

**Fuente:** Instituto de Astrofísica de Canarias (1986).

* Formación de Estructuras a Gran Escala: Las simulaciones cosmológicas muestran que la formación de las galaxias y cúmulos sólo pudieron formarse en el tiempo disponible si se considera tanto la materia bariónica como la materia oscura [2, 14].

![BOSS Great Wall](Imagenes/BOSS_Great_Wall.jpg)

**Figura 6.** Representación gráfica de un superclúster similar a la "Gran Muralla BOSS".

**Fuente:** Volker Springel - Max Planck Institute For Astrophysics (2016).

Más adelante, retomaremos el tema de las curvas de rotación.


### 1.C Hipótesis Alternativas y Estado Actual

Aunque el concenso general actual es trabajar con el modelo de la materia oscura, existen hipótesis que buscan explicar las mediciones atribuidas a la materia oscura sin la necesidad de introducir una nueva forma de materia.

Normalmente estas hipótesis se centran en la idea de que nuestra comprensión de la gravedad está incompleta y no es completamente correcta a escalas galácticas o cosmológicas [15] o bien tratan de abordarla desde otra perspectiva, como que no se trata de una interacción fundamental sino de un fenómeno relacionado con información cuántica y entropía [16]. Aún así el problema de estas hipótesis es que no logra explicar todos los fenómenos asociados a la materia oscura de manera simultánea y más bien su éxito se concentra en puntos en concreto.

Es por eso que el modelo $\Lambda$CDM (que considera materia oscura fría y energía oscura) se considera el más completo y consistente para describir las observacionales cosmológicas actuales [14, 17]. Además la evidencia observacional que respalda que el 85% de la materia del universo debe tratarse de una no visible es abrumadora [2].

Aún así, existen aún muchas preguntas relacionadas con la materia oscura. La principal siendo que no sabemos cómo se compone, pero también hay problemáticas sobre cómo podríamos estudiarla o de qué forma se encuentra en los halos de materia oscura.

## 2. Curvas de Rotación

Como se mencionó en el apartado anterior, una de las evidencias observacionales más relevantes relacionadas con la materia oscura son las curvas de rotación. Estas muestran como varía la velocidad orbital v(r) de las estrellas o del gas en función de la distancia r al centro de la galaxia [14].

![Curvas de Rotación](Imagenes/eso1709b.jpg)

**Figura 7.** Representación esquemática de la rotación del disco de las galaxias en el universo temprano y en la actualidad.

**Fuente:** ESO (2017), *eso1709b*.

### 2.A Lo que esperábamos y lo que observamos

Supongamos que observamos una galaxia con forma en espiral y quisiéramos saber cómo es su velocidad en cada punto.
Lo primero que podría saltarnos a la mente es pensarlo en un vórtice (digamos como el agua girando en un desague). Si quisiéramos hacer lo análogo con un vórtice lo que haríamos sería revisar la conservación del momento angular [19]: 

$$
v_\theta (r) = \frac{\Gamma}{2 \pi r} 
$$
Donde:
* $v_\theta$ es la velocidad tangencial.
* r es la distancia al centro.
* $\Gamma$ es la circulación. En este caso, supongamos que es constante.

No es difícil ver que la velocidad decae a medida que más nos alejamos del centro (es decir, a medida que r crece) [19].

Viéndolo ahora con las leyes de Kepler y la mecánica newtoniana, para una estrella de masa m orbitando a una distancia r, la fuerza gravitatoria proporciona la fuerza centrípeta [18]:
$$
\frac{GM(r)m}{r^2} = \frac{mv^2}{r}
$$
Donde: 
* G es la constante gravitacional: $6.67430×10^{−11} m^3 kg^{−1} s^{−2}$
* M(R) es la masa de la galaxia encerrada en el radio "r".

De aquí se sigue:
$$
v(r) = \sqrt{\frac{GM(r)}{r}}
$$

Si consideramos un r suficientemente grande, que contenga prácticamente toda la masa de la galaxia, entonces [18]:

$$
v(r) \propto r^{-1/2}
$$

Es decir, entre más nos alejamos del centro la velocidad debería decaer. Sin embargo, como podemos apreciar en la **Figura 2**, esto no es lo que observamos. La velocidad se mantiene **casi constante** [2, 10, 18]. Esto sólo tiene sentido matemáticamente si la masa sigue creciendo a medida que aumenta la distancia:
$$
M(r) \propto r
$$
Lo que nos indicaba que había materia presente que no podíamos ver y eventualmente llevó a la idea de los halos de materia oscura [1].

Eston son una estructura extensa y aproximádamente esférica que rodea las galaxias y cúmulos de galaxias. Se extienden más allá del disco visible y en una galaxia espiral puede ser varias veces mayor que el radio observable de la galaxia [14,18]. Se les considera como componentes fundamentales para explicar la estabilidad dinámica y la formación de estructuras cósmicas [2].

### 2.B Estado Actual

Las curvas de rotación son uno de los tópicos más importantes relacionados al estudio de la materia oscura y en general de la gravitación [20, 21]. De hecho se utiliza también para probar modelos alternativos al $\Lambda$CDM, como el MOND [21].

Al igual que ha pasado con toda la cosmología, el paso del tiempo y la aparición de nuevas tecnologías han facilitado que se hagan estudios cada vez más sofisticados. Las bases de datos de estos estudios son muy utilizadas en documentos científicos relacionados a gravitación, astrofísica y cosmología. A continuación hacemos mención de los más importantes y los más recientes:

* Fritz Zwicky (1933): Su estudio no se centró en las curvas de rotación, aún así es relevante por su imporancia histórica. Fue el primero en proponer la existencia de una masa invisible, llamada *dunkle Materie* (materia oscura), al analizar las velocidades de las galaxias del *Cúmulo de Coma* [11].

![Retrato de Zwicky](Imagenes/ETH-BIB-Zwicky,_Fritz_(1898-1974)-Portr_01030.tif.jpg)

**Figura 8.** Fritz Zwicky. Director de Investigación de Aerojet Engineering Corporation, Azusa, California.

**Fuente:** Anónimo (1947).

* Rubin, Ford y Thonnard (1978–1980): Probablemente la evidencia observacional más famosa de la materia oscura. Mostraton que las velocidades de rotación permanecen aproximadamente constantes a grandes radios, estableciendo definitivamente el problema de las curvas de rotación [10].

![Retrato de Rubin, Ford y Thonnard](Imagenes/Rubin_Ford_Thonnard.png)

**Figura 9.** De izquierda a derecha: Kent Ford, Norbert Thonnard y Vera Rubin en 1984.

**Accesible en:** https://www.researchgate.net/figure/Figura-2-Kent-Ford-esquerda-Norbert-Thonnard-e-Vera-Rubin-em-1984-16_fig2_389200952.

* Bosma  (1981): Encontró que la distribución del gas HI (hidrógeno neutro) está estrechamente relacionada con la distribución de masa inferida por las curvas de rotación. Impulsó el estudio detallado de los halos de materia oscura [22].

* Milgrom (1983): Propuso que las curvas de rotación podrían explicarse modificando la dinámica newtoniana en lugar de introducir materia oscura, lo que inició el principal modelo alternativo a la materia oscura (MOND) [23].

![Retrato de Mordehai Milgron](Imagenes/Milgrom_Mordechai.jpg)

**Figura 10.** Mordehai Milgrom, Físico israelí y profesor del departamento de Física de la Materia Condensada del Instituto Weizmann en Rehovot, Israel.

**Fuente:** Weizmann Institute of Science (Sin Fecha).

* Persic, Salucci y Stel (1996): Analizaron cientos de galaxias y propusieron que existe una forma universal para las curvas de rotación dependiendo de la luminosidad de la galaxia [24].

* Proyecto SPARC (2016): Sus siglas son por *Spitzer Photometry and Accurate Rotation Curves*. Recopiló curvas de rotación de alta calidad para más de 170 galaxias. Se considera la referencia moderna para estudios de materia oscura, es ampliamente utilizado en análisis bayesianos y MCMC (Markov Chain Monte Carlo) [25]. Es fundamental tanto para el modelo $\Lambda$CDM como para MOND.

![Logo del proyecto SPARC](Imagenes/SPARCcooltext1.png)

**Figura 11.** Logo del proyeco SPARC.

**Accesible en:** https://astroweb.case.edu/SPARC/ .

* Proyecto KURVS (2023): Estudió galaxias con un corrimiento al rojo (z) $\approx$ 1.5, es decir cuando el universo tenía aproximadamente un tercio de su edad actual [18]. Encontraron que la mayoría de las curvas externas siguen siendo planas o crecientes, y estimaron fracciones significativas de materia oscura en un universo temprano [26].

![Logo del proyecto KURVS](Imagenes/kurvs_logo2.jpg)

**Figura 12.** Logo del proyeco KURVS.

**Accesible en:** https://astro.dur.ac.uk/kmos/kurvs/ .

* Fuzzy Dark Matter y Curvas de Rotación de Galaxias Enanas (2023): Se trata de un trabajo a partir del catálogo de galaxias enanas "Little Things" para probar el modelo de Materia Oscura Difusa. Es interesante desde la perspectiva estadística ya que utilizó técnicas MCMC [27].

* Curva de Rotación Universal Extendida (2026): Una actualización del trabajo de Persic y Salucci utilizando datos extendidos de SPARC. Los autores construyen una Curva de Rotación Universal que permite estudiar regiones dominadas por materia oscura [28].

Como se puede ver, el tema de las curvas de rotación se mantiene vigente y se utiliza tanto para poner a prueba modelos como para estudiar la distribución de materia de las galaxias.

### 2.C Aplicación de la estadística Bayesiana

Con la aparición de nuevos y más completos catálogos también surge el problema de tener que trabajar con una cantidad cada vez mayor de datos, lo cual se suele abordar desde la estadística. Sin embargo ¿Cómo podemos trabajarlos si tenemos las consecuencias pero no las causas? Thomas Bayes se hizo esta misma pregunta [29].

En sí, podríamos describir el problema central que estudió como "Si observo los resultados de un experimento ¿Qué puedo inferir acerca de la causa que los produjo?" [29]. Esto es un problema de inferencia inversa, algo que la probabilidad clásica no respondía.

Matemáticamente, partimos de la definición de probabilidad condicional:
$$
P(A|B) = \frac{P(A \cap B)}{P(B)}
$$

y también:

$$
P(B|A) = \frac{P(A \cap B)}{P(A)}
$$

Igualando ambas expresiones:
$$
P(A \cap B) = P(A|B)P(B) = P(B|A)P(A)
$$

Obtenemos:
$$
P(A|B) = \frac{P(B|A)P(A)}{P(B)}
$$

Lo interesante aquí no viene de la ecuación en sí mismo, sino en su interpretación [30]. Supongamos que:
* A: "La hipótesis es verdadera".
* B: "Hemos observado cierta evidencia".

Entonces:
* P(A): Creencia inicial.
* P(B|A): ¿Qué tan compatible es la evidencia con la hipótesis?
* P(A|B): Creencia actualizada despúes de observar la evidencia.

Luego el conocimiento previo debe modificarse de manera proporcional a qué tan bien explica la hipótesis los datos observados [30].

De esta forma, la estadística bayesiana es un método deductivo basado en el teorema de Bayes, que permite trabajar la incertidumbre sobre un parámetro físico mediante distribuciones de probabilidad [31]. Nos ayuda a trabajar los datos cuando su incertidumbre o correlación nos es desconocida y es una de las herramientas más poderosas de la cosmología de las últimas décadas. En nuestro caso, la ecuación que utilizaríamos sería [31]:

$$
P(\theta | D) = \frac{P(D|\theta)P(\theta)}{P(D)} 
$$

donde:
* D representa los datos observacionales.
* $\theta$ es el conjunto de parámetros del modelo.
* P($\theta$) es la distribución previa (prior).
* P(D|$\theta$) es la verosimilitud (likelihood).
* P($\theta$|D) es la distribución posterior (posterior).

Si pusiéramos en contexto de las curvas de rotación, los parámetros $\theta$ podrían incluir:
* Masa del halo.
* Radio de escala.
* Concentración del halo.
* Parámetros de modelos alternativos.

#### 2.C.a Métodos MCMC y Algoritmo Goodman & Weare

Por lo general en los problemas cosmológicos y de astrofísica la distribución posterior no puede obtenerse analíticamente, es por eso que se utilizan métodos *Markov Chain Monte Carlo* (MCMC) [32].

Las cadenas de Markov (Markov Chains) son procesos aleatorios que tienen la propiedad de que, tomando en cuenta un estado actual, el futuro es independiente del pasado [32]. En otras palabras, el estado futuro depende únicamente del estado actual y no de cómo se llegó a él [32].

Un ejemplo burdo sería si quisiéramos saber si el día de mañana puede o no llover en nuestra ciudad. Estaríamos de acuerdo que si el día de hoy está soleado, lo más seguro es que al día siguiente siga soleado o a lo mucho se nuble pero sería extraño que lloviera. Si queremos saber el clima al día siguiente no necesitamos saber cómo estuvo hace un mes, sólo importa el estado actual.

![Lloro-Colombia](Imagenes/Lloro_Colombia.jpg)

**Figura 12.** Lloró, Colombia. Considerada la ciudad más lluviosa del mundo [34].

**Accesible en:** https://www.colombiaturismoweb.com/DEPARTAMENTOS/CHOCO/MUNICIPIOS/LLORO/LLORO.htm .

Las cadenas de Markov nos resultan útiles ya que podemos trabajar nuestras mediciones incluso si no conocemos los estados iniciales.

Por su parte los métodos Monte Carlo son convenientes cuando las distribuciones de probabilidad de estos eventos se vuelven demasiado complicadas para resolverlas analíticamente o por medio de nuestras computadoras [32].

Los algoritmos MCMC generan una secuencia de muestras cuya distribución converge a la distribución posterior buscada [31], es decir que en lugar de calcular una distribución difícil mejor construimos una cadena de Markov que pase más tiempo en las regiones más probables. A partir de estas se pueden estimar valores más probables de los parámetros, intérvalos de credibilidad, correlaciones, incertidumbres del modelo, etc [31].

Ejemplo: Supongamos que estamos en una ciudad dividida en cuadras y queremos ir a dónde vive la mayor cantidad de gente pero no tenemos un mapa de densidad poblacional. Nustra única herramienta es caminar por la ciudad.

Si elegimos cuadras al azar, pasaríamos muchísimo tiempo visitando zonas que no son muy relevantes. Una estrategia alternativa sería estar un tiempo en una cuadra y luego elegir una cuadra vecina, si esta parece más poblada entonces vamos pero si parece menos poblada la dejamos para después. Así pasaríamos más tiempo en las regiones más densamente pobladas. La frecuencia con la que visitamos uan región termina reflejando su probabilidad.

![Edificio Altino Arantes](Imagenes/Webysther_20190306131041_-_Edifício_Altino_Arantes.jpg)

**Figura 12.** Edificio Altino Arantes en Sao Paulo capital, Brasil. La ciudad más poblada de América [35].

**Fuente:** Webysther Nunes (2019).

Los métodos MCMC son especialmente útiles en curvas de rotación porque los moderlos de los halos suelen involucrar parámetros altamente correlacionados [31].

Existen varios tipos de algoritmos MCMC, el más popular quizás sea el Metropolis-Hastings [32], sin embargo en este trabajo nos centraremos en el algoritmo Goodman & Weare debido a su practicidad a la hora de trabajar en python, R o Julia.

Introducido por Jonathan Goodman y Jonathan Weare en 2010 en su trabajo *"Ensemble samplers with affine invariance."* [33], el algoritmo Goodman & Weare (G&W) se diferencía de altoritmos tradicionales como Metropolis-Hastings ya que utiliza un conjunto de cadenas simultáneas denominadas *walkers*. Cada *walker* explora el espacio de parámetros utilizando la información de los demás *walkers* [33]. Esto le da la ventaja de que es *invariante afín* [33], es decir, matiene su eficiencia incluso si los parámetros presentan escalas diferentes o fuertes correlaciones.

El algoritmo G&W se popularizó en astronomía y cosmología gracias a la biblioteca EMCEE, desarrollada por Daniel Foreman-Mackey y colaboradoes [36]. Esta es una imprementación en Python, estable y bien probada, del muestrador para MCMC [36]. Otros lenguajes también tienen implementaciones que trabajan con G&W u otros algoritmos MCMC. En R encontramos *LaplaceDemon* [37] la cual incluye eplícitamente dicho procedimiento, además de *Bayesian Tools* [38] y *rstan* [39] que trabajan con otros mecanismos MCMC. Para Julia (lenguaje que se ha vuelto muy popular debido a su velocidad) el equivalente a EMCEE sería *AffinelnvariantMCMC.jl* [40] aunque también tenemos *Turin.jl* [41] que actualmente es el marco bayesiano más utilizado en Julia.

El ejemplo que haremos a continuación hará uso de EMCEE en Python, pero si se tienes curiosidad puedes intentar utilizar R o Julia desde google colab, ya que soporta los 3 entornos, e intentar replicar lo obtenido.

## 3 Ejemplo de aplicación

Finalmente, veamos un poco de cómo funciona EMCEE en Python. Vamos a ver paso a paso un algoritmo que utilice esta implementación, revisaremos cuál es la función de cada parte del código y, a partir de dos conjuntos de datos del proyecto SPARC [25] de galaxias diferentes, a la par de irlo ejecutando también veremos cómo se obtiene la densidad característica, el radio de escala de un halo de materia oscura y a qué se refieren conceptos como el "likelihood" o los "priors".

### 3.A Librerías

Siempre al comenzar debemos colocar las librerías e implementaciones que vamos a necesitar. En este caso utilizaremos:

* pandas
* numpy
* scipy
* matplotlib
* emcee
* corner
* tqdm

Como no todas vienen por defecto en el entorno de python que estamos usando, vamos a instalarlas.

In [3]:
#Ejecuta esta celda primero
!mamba install pandas numpy scipy matplotlib emcee corner

mambajs 0.19.13

Specs: xeus-python, numpy, matplotlib, pillow, ipywidgets>=8.1.6, ipyleaflet, scipy, pandas, emcee, corner
Channels: emscripten-forge, conda-forge

Solving environment...
Solving took 0.7393999999761581 seconds
All requested packages already installed.


In [4]:
# Luego ejecuta esta
!mamba install tqdm

mambajs 0.19.13

Specs: xeus-python, numpy, matplotlib, pillow, ipywidgets>=8.1.6, ipyleaflet, scipy, pandas, emcee, corner, tqdm
Channels: emscripten-forge, conda-forge

Solving environment...
Solving took 0.7315 seconds
  Name                          Version                       Build                         Channel                       
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
+ __unix                        1.0.0                         hc286ada_0                    emscripten-forge              
+ tqdm                          4.67.3                        pyh8f84b5b_0                  conda-forge                   


In [6]:
#Y finalmente esta para importar las librerías e implementaciones que acabamos de instalar.

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from scipy.integrate import quad
from scipy.optimize import minimize

import emcee
import corner

### 3.B Los Datos

En este caso, vamos a trabajar con un set de datos del proyecto SPARC [25] obtenidos del repositorio en línea de *The Data Behind Dark Matter: Exploring Galactic Rotation* [42].

Al ser un catálogo de galaxias bastante extenso (175) [25] haremos elección solamente de dos para fines académicos. Estas son:
* NGC 3198: También conocida como Herschel 146, se trata de una galaxia masiva de forma de espiral que presenta una curva clásica limpia [43]. La tomaremos de referencia ya que es un ejemplo clásico de este problema.

![NGC3198](Imagenes/NGC3198.jpg)

**Figura 13.** Galaxia NGC 3198.

**Fuente:** KPNO/NOIRLab/NSF/AURA (2020). *iotw2035a*.
  
* DDO 161: Una galaxia enana de baja masa y dominada por materia oscura [44]. Resulta interesante ya que son objetos menos estudiados y se usa como una prueba del modelo $\Lambda$CDM a "pequeñas escalas" [44].

![DDO161](Imagenes/DDO161.jpg)

**Figura 13.** Galaxia DDO 161.

**Fuente:** NOIRLab/NSF/AURA/P. Horálek (Institute of Physics in Opava) (2025). *iotw2547a*.

Tener esta contraposición de galaxias será interesante para ver cómo se comporta nuestro modelo en ambas y poder comparar resultados entre ellas.

Los conjuntos de datos que proporciona SPARC contienen:
* Rad -> radio galactocéntrico
* Vobs -> velocidad observada
* errV -> error observacional
* Vgas -> contribución del gas
* Vdisk -> contribución del disco estelar
* Vbul -> contribución del bulbo
* SBdisk -> brillo superficial del disco
* SBbul -> brillo superficial del bulbo

A partir de estos datos, trabajaremos con estadística bayesiana sobre el modelo.

Para importar nuestros datos usaremos "import". Debido al formato que tienen es importante señalar que las partes con "#" son comentarios para que no las lea y no tengamos problemas más adelante.

In [14]:
#Ejecuta esta celda para importar nuestros datos como "data".
data = np.loadtxt(
    "Data/NGC3198_rotmod.dat",
    comments="#"
)

#Luego esta parte extrae las columnas y define la serie de datos
Rad    = data[:,0]
Vobs   = data[:,1]
errV   = data[:,2]
Vgas   = data[:,3]
Vdisk  = data[:,4]
Vbul   = data[:,5]
SBdisk = data[:,6]
SBbul  = data[:,7]

### 3.C El Modelo

Se refiere a la descripción matemática de nuestro problema. En este caso ya está definida en la literatura, así que partiremos de la relación física que se usa en curvas de rotación [45].

$$
V_{obs}^{2}(r) = V_{gas}^{2}(r) + V_{disk}^{2}(r) + V_{bul}^{2}(r) + V_{DM}^{2}(r)
$$

Para facilidad consideraremos:
$$
V_{DM} = V_{h} = cte.
$$

Sustituyendo esto en la ecucación, renombrando $V_{obs}$ como $V_{mod}$ y despejando para $V_{mod}$ se obtiene:

$$
V_{mod} = \sqrt{V_{gas}^{2}(r) + V_{disk}^{2}(r) + V_{bul}^{2}(r) + V_{h}^{2}}
$$

E intentaremos encontrar $V_h$ usando EMCEE. Cabe mencionar que este modelo es poco riguroso trabajarlo así ya que estamos fijando varios parámetros y trabajando únicamente sobre uno (que además estamos considerando constante). Aún así para fines académicos es útil para ilustrar cómo funciona un algoritmo de este tipo.

Si se quisiera hacer algo más completo se podría trabajar con la siguiente expresión:

$$
V_{obs}^{2}(r) = V_{gas}^{2}(r) + \Upsilon_d V_{disk}^{2}(r) + \Upsilon_b V_{bul}^{2}(r) + V_{DM}^{2}(r)
$$

Donde: 
* $\Upsilon_d$ es la razón masa-luz del disco.
* $\Upsilon_b$ es la razón masa-luz del bulbo.
* $V_DM$ es la contribución del halo de materia oscura.

Se podría trabajar con estos parámetros libres o incluso fijar dos de ellos (digamos, $\Upsilon_d$ y $\Upsilon_b$) pero aumenta un poco la complejidad del problema. Para mantenerlo como algo meramente ilustrativo de MCMC vamos a fijar toda la parte bariónica y trabajar sólo con el parámetro $V_h$.

Ahora definiremos nuestro modelo en Python.

In [9]:
#Ejecuta esta celda para definir nuestro modelo como una función. Su nombre será "model_velocity".
#Depende de los parámetros vh, vgas, vdisk y vbul. En otras palabras, es la ecuación que usamos para V_mod.
def model_velocity(Vh,
                   Vgas,
                   Vdisk,
                   Vbul):

    return np.sqrt(
        Vgas**2 +
        Vdisk**2 +
        Vbul**2 +
        Vh**2
    )

También es importante definir nuestro vector de parámetros $\theta$. Como ya mencionamos anteriormente, sólo nos interesa un parámetro, así que:
$$
\theta = V_h
$$

In [10]:
#Ejecuta esta celda.
ndim = 1

### 3.D El Prior

El prior (o distribución a priori) representa lo que sabemos o asumismos sobre los parámetros antes de incorporar los datos observacionales [46]. Matemáticamente se escribe:

$$
P(\theta)
$$

En nuestro caso debemos imponer que la velocidad sea no-negativa, pero debemos considerar un límite máximo ya que si lo dejamos abierto tendremos problemas más adelante con el likelihood.

Una buena regla sería asumir un valor máximo a partir de la velocidad máxima observada [46]. En este caso, de nuestro conjunto de datos [42], podemos ver que la velocidad observada más alta fue de 157 km/s. Así podríamos asumir un límite de 300 km/s (al rededor del doble).

$$
0 < V_h < 300
$$

Entonces:

$$
P(V_h)=
\begin{cases}
1, & 0 < V_h < 300 \\
0, & \text{otro caso}
\end{cases}
$$

Como EMCEE trabaja con logaritmos [36], nosotros trabajaremos con el logaritmo del prior para facilitar los cálculos.

In [12]:
#Ejecuta esta celda, esta define nuestra función del logaritmo del prior "log_prior".
def log_prior(theta):

    Vh = theta[0]

    if 0 < Vh < 300:
        return 0.0

    return -np.inf

### 3.E Construir la función Chi-Cuadrada ($\chi^2$)

Para poder comprar la observación con el modelo hacemos uso de la función *Chi-cuadrada* ($\chi^2$) [46] mediante:

$$
\chi^2 = \sum_i{\frac{(v_{obs,i}-v_{mod,i})^2}{\sigma^2_i}}
$$

idealmente $v_{obs,i}=v_{mod,i}$, es decir que cada velocidad del modelo es exactamente igual a la observada. Sin embargo esto no sucedería a menos que tuviéramos un modelo 100% preciso (algo matemáticamente improbable).

In [13]:
#Ejecuta esta celda para definir la función chi-cuadrada como "chi2"
def chi2(theta,
         Vobs,
         errV,
         Vgas,
         Vdisk,
         Vbul):

    Vh = theta[0]

    Vmodel = model_velocity(
        Vh,
        Vgas,
        Vdisk,
        Vbul
    )

    return np.sum(
        ((Vobs - Vmodel)/errV)**2
    )

### 3.F Likelihood

La likelihood (verosimilitud) mide que tan bien un conjunto de parámetros reproduce los datos observados [46]. Se define como:

$$
P(D|\theta)
$$

y se interpreta como "la probabilidad de observar los datos D suponiendo que los parámetros son $\theta$".

Se espera que el comportamiento de nuestros errores sea gaussiano [36] así que:

$$
\mathcal{L} \propto e^{-\chi^2/2}
$$

donde $\mathcal{L}$ es la likelihood. Es fácil ver que luego con el logaritmo se sigue:

$$
\ln \mathcal{L} = -\frac{\chi^2}{2}
$$

Ahora vamos con el código:

In [16]:
#Ejecuta esta celda. Definimos el logaritmo del likelihood como log_likelihood.
def log_likelihood(theta,
                   Vobs,
                   errV,
                   Vgas,
                   Vdisk,
                   Vbul):

    return -0.5 * chi2(
        theta,
        Vobs,
        errV,
        Vgas,
        Vdisk,
        Vbul
    )

### 3.G La Posterior

La posterior es el resultado final del análisis bayesiano [46], combina la información previa (prior) con la información contenida en los datos (likelihood) y se obtiene a partir del teorema de Bayes:

$$
P(\theta | D) = \frac{P(D|\theta)P(\theta)}{D}
$$

Donde:
* $P(\theta|D)$ es la posterior.
* $P(D|\theta)$ es la likelihood.
* $P(\theta)$ es la prior.
* $P(D)$ es la evidencia bayesiana.

Pero como EMCEE no necesita calcular la evidencia [36], normalmente se usa:

$$
P(\theta|D) \propto P(D|\theta) P(\theta)
$$

Ahora bien, al trabajar con logaritmos también es más sencilla esta parte, ya que si aplicamos el logaritmo a ambas partes de la igualdad obtenemos:

$$
\ln(P(\theta|D)) \propto \ln (P(D|\theta) P(\theta)) = \ln (P(D|\theta)) + \ln (P(\theta))
$$

Y como recordaremos $\ln (P(D|\theta))$ es el logaritmo del likelihood y $\ln (P(\theta))$ el del prior. Ahí está otra ventaja de trabajar sobre los logaritmos en vez de sobre las funciones en sí.

In [17]:
#Ejecuta esta celda para definir el logaritmo del posterior. Lo llamaremos log_probability.
def log_probability(theta,
                    Vobs,
                    errV,
                    Vgas,
                    Vdisk,
                    Vbul):

    lp = log_prior(theta)

    if not np.isfinite(lp):
        return -np.inf

    return (
        lp +
        log_likelihood(
            theta,
            Vobs,
            errV,
            Vgas,
            Vdisk,
            Vbul
        )
    )

### 3.H Walkers y EMCEE

Lo último que nos faltaría para poder ejecutar EMCEE es inicializar a nuestros *walkers*, que como ya dijimos nos ayudarán a explorar el espacio de parámetros [36]. Para esto hay que definir la cantidad de walkers y su punto de partida.

Para la cantidad de walkers, lo que se sugiere [36] es utilizar 2 walkers por cada parámetro del vector de parámetros. Sin embargo, ya que en nuestro caso es 1, tomaremos un número más grande. Es importante mencionar que entre más walkers se utilice menor diferencia hay (en tanto superen la proporción de 2 a 1) [36] y sólo hará más lento el proceso. En este caso consideraremos usar 32 walkers, aunque siéntete libre de modificar el código para aumentar o disminuir walkers y ver qué sucede.

En cuanto al punto inicial, lo que requerimos es que inicie dentro del espacio de parámetros [36]. En este caso, vamos a colocarlo en 50 km/s aunque igualmente puedes modificar este punto inicial y ver qué pasa.

In [18]:
#Ejecuta esta celda para inicializar los walkers.

#Cambia este valor para aumentar o disminuir el número de walkers. Por defecto usamos 32 en este ejemplo.
nwalkers = 32


# Cambia este valor para modificar el punto de inicio. Considera que debe ser no negativo y menor a 300. Por defecto usamos 50 en este ejemplo.
initial = 50

pos = (
    initial
    + 1e-2*np.random.randn(
        nwalkers,
        1
    )
)

Ahora si, con todo listo vamos a ejecutar EMCEE.

In [21]:
#Ejecuta esta celda para iniciar EMCEE. Puedes ejecutarla varias veces y ver cómo cambia el resultado.
#Tras ejecutar la celda, espera un poco a que aparezca el mensaje "✓ Muestreo completado".

sampler = emcee.EnsembleSampler(
    nwalkers,
    ndim,
    log_probability,
    args=(
        Vobs,
        errV,
        Vgas,
        Vdisk,
        Vbul
    )
)

#Si quieres ver todos los datos que arroja y el porcentaje de avance, puedes cambiar el "False" luego de "progress" por "True" y ejecutar
#la celda.
#Si lo haces, ignora los mensajes de error y espera a que termine. Si luego de haberlo hecho quieres que no se vean los datos, vuelve a 
#escribir "False" y ejecuta la celda.

sampler.run_mcmc(
    pos,
    5000,
    progress=False
)

print("✓ Muestreo completado")

✓ Muestreo completado


### 3.I Burn-in y Resultados

Antes de pasar a revisar los resultados, tenemos que eliminar el Burn-In.

En las primeras etapas del MCMC, los walkers aún no alcanzan la distribución estacionaria (la más probable). Esta etapa se conoce como el Burn-in y por lo general debe ser ignorada [36].

Para indicar que la ignore, vamos a ejecutar la siguiente celda:

In [22]:
#Ejecuta esta celda para limpair el Burn-In.

samples = sampler.get_chain(
    discard=1000,
    flat=True
)

Ahora si, ya podemos mostrar el parámetro estimado.

Este se obtiene usando percentiles de la distribución posterior, típicamente los percentiles 16 (aproximadamente a -$1\sigma$), 50 (la mediana) y 84 (aproximadamente a +$1\sigma$), correspondientes al valor central y a las incertidumbres de $1\sigma$ [46].

In [24]:
#Ejecuta esta celda para obtener el resultado.

Vh_median = np.percentile(
    samples[:,0],
    50
)

Vh_low = np.percentile(
    samples[:,0],
    16
)

Vh_high = np.percentile(
    samples[:,0],
    84
)

print(f"Vh = {Vh_median:.2f} km/s")
print(f"Percentil 16 = {Vh_low:.2f} km/s")
print(f"Percentil 84 = {Vh_high:.2f} km/s")

Vh = 85.81 km/s
Percentil 16 = 85.26 km/s
Percentil 84 = 86.35 km/s


Ahora bien ¿Este ajuste es bueno? Para esto vamos a compararlo con lo que el modelo nos indica.

Primero debemos definir nuestra función que será el "modelo".

In [26]:
#Ejecuta esta celda, nos dará el modelo.
Vmodel = np.sqrt(
    Vgas**2 +
    Vdisk**2 +
    Vbul**2 +
    Vh_median**2
)

Luego calculamos el mejor ajuste de $\chi^2$[46]:

In [28]:
#Ejecuta esta celda, comparará las v_obs con las v_model.
chi2_best = np.sum(
    ((Vobs - Vmodel)/errV)**2
)

print(f"Chi² = {chi2_best:.3f}")

Chi² = 4704.038


Calculamos el $\chi^2$ reducido utilizando [46]:
$$
\chi^2_{\nu} = \frac{\chi^2}{N-p}
$$

donde:
* N es el número de observaciones.
* p es el número de parámetros libres.


In [29]:
#Ejecuta esta celda. Calcula la chi cuadrada reducida.
N = len(Vobs)
p = 1

chi2_red = chi2_best/(N-p)

print(f"Chi² reducido = {chi2_red:.3f}")

Chi² reducido = 112.001


Por lo general lo que querríamos sería un valor lo más cercano a 1 posible, ya que esto indicaría un buen ajuste. Si el resultado es mucho menor a 1 entonces los errores están sobreestimados, mientras que si es mayor a 2 el modelo es ineficiente [46].

Ahora bien, en nuestro caso es muchísimo mayor a 2 pero es de esperarse. Si recordamos, decidimos sobresimplificarlo para fines del ejercicio y es por eso que nuestro modelo no representa muy bien la realidad.

Podemos también revisar el parámetro estimado y el rango de velocidades:

In [33]:
#Ejecuta esta celda, veremos los 3 parámetros: el estimado, el mínimo de nuestro conjunto de datos y el máximo.
print("Valor estimado =", Vh_median)
print("Valor mínimo =", np.min(Vobs))
print("Valor máximo =", np.max(Vobs))

Valor estimado = 85.80669666322555
Valor mínimo = 24.4
Valor máximo = 157.0


Vemos que de hecho el valor estimado está dentro del rango y cerca del punto medio del rango de datos, esto nos dice que nuestra estimación no fue tan exagerada pero el valor de $\chi^2_\nu$ nos indica que (aunque probablmenete nuestro valor sea aceptable) la manera en la que describimos el problema se aleja mucho de la realidad (lo cual, como ya habíamos dicho, es comprensible).

### 3.J Reutilizar EMCEE

Luego de todo esto uno podría preguntarse "¿Entonces cada vez que quiera usar EMCEE en el mismo notebook tengo que definirlo una y otra vez?" Y en términos generales la respuesta es no. Siempre que el problema que abordas sea similar, puedes seguir utilizando todas las funciones que ya definimos y únicamente centrarse en importar los datos de manera correcta. Eso sí: Si se quiere que no se sobre-escriban los valores ya calculados o importados, hay que modificar algunas etiquetas.

Para mostrar cómo, ahora haremos EMCEE para la segunda galaxia que mencionamos (DDO 161).

Comenzamos importando los datos.

In [34]:
#Ejecuta esta celda para importar nuestros datos como "data2".
data2 = np.loadtxt(
    "Data/DDO161_rotmod.dat",
    comments="#"
)

#Luego esta parte extrae las columnas y define la serie de datos. Nótese que ahora debemos de tener
#cuidado de poner "2" después de cada dato y especificar que es el set de datos "data2" para no
#sobreescribir los datos que teníamos anteriormente y que se importen los de la segunda galaxia.
Rad2    = data2[:,0]
Vobs2   = data2[:,1]
errV2   = data2[:,2]
Vgas2   = data2[:,3]
Vdisk2  = data2[:,4]
Vbul2   = data2[:,5]
SBdisk2 = data2[:,6]
SBbul2  = data2[:,7]

Como tanto el modelo como los priors, likelihood, posteriors y $\chi^2$ son las mismas, entonces ya podríamos ejecutar los walkers y el EMCEE.

Ahora bien, si queremos hacerlo reutilizable conviene hacer un pequeño cambio y en vez de definir cada parte del proceso por separado ponerlo todo junto en una misma función. Esto no sólo nos ahorra espacio de trabajo sino que ahora cada vez que queramos usarlo solo tenemos que llamar a la función.

Así:

In [35]:
#Ejecuta esta celda para definir nuestro proceso EMCEE como run_EMCEE.
#Vemos que ahora el número de walkers, de pasos, punto de inicio y el burn-in lo definimos como
#variables de la función.
def run_emcee(
    Vobs,
    errV,
    Vgas,
    Vdisk,
    Vbul,
    nwalkers=32,
    nsteps=5000,
    burnin=1000,
    initial=50
):
#Aquí vemos la dimensión de nuestro vector de parámetros.
    ndim = 1

#Posiciones iniciales de los walkers. Tiene una pequeña perturbación aleatoria para evitar que
#ocupen el mismo punto. 
    pos = (
        initial
        + 1e-2*np.random.randn(
            nwalkers,
            ndim
        )
    )
#Creación del sampler (muestreador), aquí es donde el número de walkers entra.
    sampler = emcee.EnsembleSampler(
        nwalkers,
        ndim,
        log_probability,
        args=(
            Vobs,
            errV,
            Vgas,
            Vdisk,
            Vbul
        )
    )
#Se ejecuta el algorimo MCMC.
    sampler.run_mcmc(
        pos,
        nsteps,
        progress=False
    )
#Se descarta el Burn-In y se obtienen las muestras.
    samples = sampler.get_chain(
        discard=burnin,
        flat=True
    )
#Se estima Vh.
    Vh_median = np.percentile(
        samples[:,0],
        50
    )
# Límite inferior del intervalo de credibilidad (percentil 16).
    Vh_low = np.percentile(
        samples[:,0],
        16
    )
# Límite superior del intervalo de credibilidad (percentil 84).
    Vh_high = np.percentile(
        samples[:,0],
        84
    )

    return {
        "sampler": sampler,
        "samples": samples,
        "Vh": Vh_median,
        "Vh_low": Vh_low,
        "Vh_high": Vh_high
    }

Y ahora si queremos ejecutarlo, sólo falta ejecutar esta función con los datos de la galaxia que nos interesa.

In [37]:
#Ejecuta esta celda para encontrar Vh para la segunda galaxia.
resultado2 = run_emcee(
    Vobs2,
    errV2,
    Vgas2,
    Vdisk2,
    Vbul2
)

print(resultado2["Vh"])

49.27515841450506


Podemos ejecutarlo también para la primera nuevamente (el resultado cambiará un poco por la aleatoriedad pero estará dentro del rango de los percentiles que obtuvimos anteriormente.

In [39]:
#Ejecuta esta celda para encontrar Vh para la primer galaxia.
resultado1 = run_emcee(
    Vobs,
    errV,
    Vgas,
    Vdisk,
    Vbul
)

print(resultado1["Vh"])

85.8112093036882


Así podríamos obtener valores para distintas galaxias simplemente importando los datos y ejecutando la función que definimos.

## 4. Conclusiones

Podemos ver que el concepto de la materia oscura es al mismo tiempo útil y misterioso. A pesar de haberlo estudiado por tanto tiempo, aún se mantiene como un objeto del que sabemos poco. Y aunque hay hipótesis alternativas, el modelo de $\lambda$CDM se sigue manteniendo como el mejor hasta el momento, sobreviviendo a las pruebas el paso de los años.

Las curvas de rotación son facinantes y un buen método para probar diferentes modelos cosmológicos. A medida que nuestra tecnología avanza también podemos tener catálogos más extensos y con eso hacer pruebas que afinen nuestros modelos. Es un tema que se mantiene vigente y hay mucha literatura al respecto.

La estadística bayesiana es un método de inferencia inversa muy útil en el campo de la cosmología y la astrofísica donde no podemos replicar experimentos y sólo podemos trabajar con lo que observamos. Los métodos MCMC son muy útiles para trabajar computacionalmente estos datos y, aunque el estándar sigue siendo Python, aún hay alternativas en otros lenguajes como R y Julia que pueden ser útiles dependiendo de la perspectiva que se le de al problema que se quiera estudiar.

## 5. Cuestionario

Después de leer el notebook, contesta las siguientes preguntas para repasar lo visto.

1. ¿Cómo podemos definir la materia oscura?
   1. Materia que sólo interactúa consigo misma.
   2. Materia que no interactúa gravitacionalmente pero emite grandes cantidades de radiación electromagnética.
   3. Materia que interactía gravitacionalmente pero no emite cantidades detectables de radiación electromagnética.
   4. No hay una definición que podamos usar.
2. Al resto de la materia (no oscura) se le conoce como:
   1. Materia Bariónica.
   2. Materia común.
   3. Materia Bosónica.
   4. Materia visible.
3. Verdadero o Falso: Aproximádamente el 85% de la materia que existe en el universo es oscura.
    1. Verdadero.
    2. Falso.
4. Son evidencias observacionales de la materia oscura.
    1. Curvas de rotación, dinámica de cúmulos de galaxia, lentes gravitacionales, formación de estructuras a gran escala.
    2. Curvas de rotación, dinámica newtoniana, ley de Kepler, formación de estructuras a gran escala.
    3. Curvas de rotación, agujeros negros, los huracanes, formación de estructuras a gran escala.
    4. Curvas de rotación, el superclúster Gran Muralla BOSS, DDO 161, NGC 3198.
    5. No hay evidencia observacional. 
5. ¿Qué crítica se le puede hacer a las hipótesis alternativas a la materia oscura?
    1. Que no saben de física.
    2. Que funciona mejor el método EMCEE.
    3. Que por lo general solucionan un problema concreto pero no todos en su conjunto.
    4. Ninguna porque es lo que más se usa en cosmología.
6. Verdadero o Falso: Las curvas de rotación pueden explicarse considerando sólo materia bariónica.
    1. Verdadero.
    2. Falso.
7. ¿Qué son los halos de materia oscura?
    1. No sabemos.
    2. Un videojuego creado por Bungie.
    3. Estructuras extensas y esféricas que rodean las galaxias y cúmulos de galaxias.
    4. Estructuras extensas y cúbicas que rodean las galaxias y cúmulos de galaxias
8. ¿Quién fue el primero en introducir el término "Materia Oscura"?
    1. Boltzmann.
    2. Schrödinger.
    3. Zwicky.
    4. Newton.
9. Establecieron el problema de las Curvas de Rotación.
    1. Newton y Leibniz.
    2. Emerson, Lake y Palmer.
    3. Einstein, Schrödinger y Pauli.
    4. Rubin, Ford y Thonnard
10. Es el principal modelo alternativo al $\lambda$CDM.
    1. DOI161.
    2. MOND.
    3. KURVS.
    4. Little THINGS.
11. De este proyecto proviene el catálogo más utilizado para estudios de materia oscura.
    1. KURVS.
    2. Superclúster Gran Muralla BOSS.
    3. SPARC.
    4. NGC 3198.
12. Verdadero o Falso: La principal contribución de Thomas Bayes fue la interpretación de su fórmula.
    1. Verdadero.
    2. Falso.
13. En los eventos de tipo Cadena de Markov, se considera que el estado futuro depende de:
    1. Todos los puntos anteriores menos el actual.
    2. Solo el estado actual.
    3. No se puede determinar su dependencia.
    4. Todos los puntos que existen.
14. ¿Cuando es conveniente usar métodos Monte Carlo?
    1. Cuando las distribuciones de probabilidad se vuelven demasiado complicadas.
    2. Cuando las distribuciones de probabilidad son guassianas.
    3. Cuando las distribuciones de probabilidad no existen.
    4. No sirven para nada.
15. Lo que caracteriza al algoritmo Goodman & Weare es:
    1. Los Walkers.
    2. Que es el único algoritmo MCMC que se utiliza en cosmología moderna.
    3. Que es el único algoritmo MCMC que tiene una librería en Python.
    4. No tiene nada de especial.
16. Es el leguanje en el que trabaja la librería EMCEE.
    1. Python.
    2. R.
    3. Julia.
    4. KURVS.
17. ¿Cuál de estas opciones describe mejor lo que es un Prior?
    1.  Representa lo que sabemos o asumismos *a priori*.
    2.  Es un tipo de algoritmo MCMC.
    3.  Correlaciona lo que asumimos con los datos que observamos y su distribución.
    4.  Es equivalente al Likelihood.
18. Es una de las ventajas de trabajar con los logaritmos de Priors, Posteriors y Likelihood.
    1. Se puede definir $\ln (Posteriors) = \ln (Priors) + \ln (Likelihood)$.
    2. Siempre cuadra el resultado.
    3. Se puede trabajar en R, Python y Julia.
    4. No tiene una ventaja significativa.
19. Verdadero o Falso: Una vez ejecutado EMCEE se pueden obtener las estimaciones.
    1. Verdadero.
    2. Falso.
20. Verdadero o Falso: Si se quiere usar varias veces EMCEE en un mismo documento en python es necesario definirlo por completo cada vez.
    1. Verdadero.
    2. Falso.



## 6. Bibliografía.
1. Vázquez-González, A., & Matos, T. (2008). La materia oscura del universo: retos y perspectivas. Revista mexicana de física E, 54(2), 193-202.
2. Casadiego, R. P. (2024). ¿ Qué es la Materia Oscura?: El Enigma de la Materia Oscura Respondido a Través del Plasma Cristalizado. rogelio perez.
3. Sean Carroll, S. (2019). Spacetime and geometry: An introduction to general relativity (2nd ed.). Cambridge University Press.
4. European Space Agency. (2013). Planck cosmic recipe. ESA. https://www.esa.int/ESA_Multimedia/Images/2013/03/Planck_cosmic_recipe
5. NASA. (2013). Refining the ingredients of our universe. NASA. https://science.nasa.gov/photojournal/refining-the-ingredients-of-our-universe/
6. Peter, A. H. G. (2012). Dark Matter: A Brief Review. arXiv. https://arxiv.org/abs/1201.3942
7. Bahcall, N. A. (2015). Dark matter universe. Proceedings of the National Academy of Sciences, 112(40), 12243–12245. https://doi.org/10.1073/pnas.1516944112
8. Misiaszek, M., & Rossi, N. (2024). Direct Detection of Dark Matter: A Critical Review. Symmetry, 16(2), 201. https://doi.org/10.3390/sym16020201
9. U.S. Department of Energy. (2016). The Search for Dark Matter. U.S. Department of Energy. https://www.energy.gov/science/articles/search-dark-matter
10. Rubin, V. C., Ford, W. K., Jr., & Thonnard, N. (1980). Rotational properties of 21 SC galaxies with a large range of luminosities and radii. The Astrophysical Journal, 238, 471–487. https://doi.org/10.1086/158003
11. Zwicky, F. (1933). Die Rotverschiebung von extragalaktischen Nebeln. Helvetica Physica Acta, 6, 110–127.
12. Clowe, D., Bradac, M., Gonzalez, A. H., Markevitch, M., Randall, S. W., Jones, C., & Zaritsky, D. (2006). A direct empirical proof of the existence of dark matter. The Astrophysical Journal Letters, 648(2), L109–L113. https://doi.org/10.1086/508162
13. Planck Collaboration. (2020). Planck 2018 results. VI. Cosmological parameters. Astronomy & Astrophysics, 641, A6. https://doi.org/10.1051/0004-6361/201833910
14. Barbara Ryden. (2017). Introduction to Cosmology (2nd ed.). Cambridge University Press.
15. Milgrom, M. (1983). A modification of the Newtonian dynamics as a possible alternative to the hidden mass hypothesis. The Astrophysical Journal, 270, 365–370.
16. Verlinde, E. (2017). Emergent Gravity and the Dark Universe. SciPost Physics, 2(3), 016.
17. Scott Dodelson, S., & Schmidt, F. (2020). Modern Cosmology (2nd ed.). Academic Press.
18. Liddle, A. (2015). An Introduction to Modern Cosmology (3rd ed.). Wiley. ISBN 978-1-118-50714-8.
19. White, F. M. (2016). Fluid Mechanics (8th ed.). McGraw–Hill Education.
20. James Binney, J., & Scott Tremaine, S. (2008). Galactic Dynamics (2nd ed.). Princeton University Press.
21. Stacy McGaugh. (2020). Predictions and Outcomes for the Dynamics of Rotating Galaxies. Galaxies, 8(2), 35. https://doi.org/10.3390/galaxies8020035
22. Albert Bosma. (1981). 21-cm line studies of spiral galaxies. Astronomical Journal, 86, 1825–1846.
23. Mordehai Milgrom. (1983). A modification of the Newtonian dynamics as a possible alternative to the hidden mass hypothesis. The Astrophysical Journal, 270, 365–370. https://doi.org/10.1086/16113
24. Massimo Persic, M., Paolo Salucci, P., & Stel, F. (1996). The Universal Rotation Curve of Spiral Galaxies. Monthly Notices of the Royal Astronomical Society, 281(1), 27–47.
25. Federico Lelli, Stacy McGaugh, & Schombert, J. M. (2016). SPARC: Mass Models for 175 Disk Galaxies with Spitzer Photometry and Accurate Rotation Curves. The Astronomical Journal, 152(6), 157. https://doi.org/10.3847/0004-6256/152/6/157
26. Puglisi, A., Dudzevičiūtė, U., Swinbank, A. M., et al. (2023). KURVS: The outer rotation curve shapes and dark matter fractions of z ∼ 1.5 star-forming galaxies. Monthly Notices of the Royal Astronomical Society, 525, 3985–4004.
27. Bañares-Hernández, A., Castillo, A., Camalich, J. M., & Iorio, G. (2023). Confronting fuzzy dark matter with the rotation curves of nearby dwarf irregular galaxies. arXiv:2304.05793.
28. Bhatia, E., Paolo Salucci, P., Schiavone, T., & Haridasu, S. (2026). Extended Universal Rotational Curve of Spiral Galaxies. arXiv:2604.27320.
29. Bayes, T., & Price, R. (1763). An Essay towards solving a Problem in the Doctrine of Chances. Philosophical Transactions of the Royal Society of London, 53, 370–418.
30. Edwin Jaynes. (2003). Probability Theory: The Logic of Science. Cambridge University Press.
31. Trotta, R. (2008). Bayes in the sky: Bayesian inference and model selection in cosmology. Contemporary Physics, 49(2), 71–104. https://doi.org/10.1080/00107510802066753
32. Grimmett, G. and Stirzaker, D. (2024) Probability and random processes Geoffrey R. Grimmett (Statistical Laboratory, Cambridge University), and David R. Stirzaker (Mathematical Institute, Oxford University). Oxford: Oxford University Press.
33. Goodman, J., & Weare, J. (2010). Ensemble samplers with affine invariance. Communications in Applied Mathematics and Computational Science, 5(1), 65–80. https://doi.org/10.2140/camcos.2010.5.65
34. NOAA (agosto de 2008). «Global Measured Extremes of Temperature and Precipitation» (en inglés). Consultado el 13 de agosto de 2012.
35. Ranking de las 30 ciudades más pobladas de américa - misceláneas - datasets globales (no date) Misceláneas - Datasets globales | Dateas.com. Available at: https://www.dateas.com/es/explore/ranking-30-ciudades-mas-pobladas-america?sortby=&order=&ft=on&mainfield=&hidVA=0&adva_all=&adva_lit=&adva_any=&adva_non= (Accessed: 31 May 2026).
36. Foreman-Mackey, D., Hogg, D. W., Lang, D., & Goodman, J. (2013). emcee: The MCMC Hammer. Publications of the Astronomical Society of the Pacific, 125(925), 306–312. https://doi.org/10.1086/670067
37. Michael Clark. (2025). LaplacesDemon: Complete Environment for Bayesian Inference. R package version 16.1.6.
38. Hartig, F., Minunno, F., & Paul, S. (2019). BayesianTools: General-purpose MCMC and SMC samplers and tools for Bayesian statistics.
39. Stan Development Team. (2025). RStan: the R interface to Stan.
40. Jonathan Goodman, J., & Jonathan Weare, J. (2010). Ensemble samplers with affine invariance. Communications in Applied Mathematics and Computational Science, 5(1), 65–80.
41. Hong Ge, H., Xu, K., & Ghahramani, Z. (2018). Turing: A language for flexible probabilistic inference. Proceedings of AISTATS.
42. Villano, A. n. et al. (2023) The data behind Dark Matter: Exploring Galactic Rotation, Journal of Open Source Education. Available at: https://jose.theoj.org/papers/10.21105/jose.00184 (Accessed: 31 May 2026).
43. Begeman, K. G. (1989). HI rotation curves of spiral galaxies. I-NGC 3198. Astronomy and Astrophysics (ISSN 0004-6361), vol. 223, no. 1-2, Oct. 1989, p. 47-60. Research supported by ZWO and ASTRON., 223, 47-60.
44. Li, J., Greene, J. E., Danieli, S., Carlsten, S. G., & Geha, M. (2026). A Possible “Too-many-satellites” Problem in the Isolated Dwarf Galaxy DDO 161. The Astrophysical Journal Letters, 998(1), L24.
45. Binney, J., & Tremaine, S. (2008). Galactic Dynamics (2nd ed.). Princeton University Press.
46. Sivia, D. S., & Skilling, J. (2006). Data analysis: A Bayesian tutorial (2nd ed.). Oxford University Press.